crypto/dh/dh_key.c

```diff
@@ -327,11 +327,13 @@ static int generate_key(DH *dh)
                 goto err;
 #else
             if (dh->params.q == NULL) {
-                /* secret exponent length, must satisfy 2^(l-1) <= p */
-                if (dh->length != 0
-                    && dh->length >= BN_num_bits(dh->params.p))
+                /* secret exponent length, must satisfy 2^l < (p-1)/2 */
+                l = BN_num_bits(dh->params.p);
+                if (dh->length >= l)
                     goto err;
-                l = dh->length ? dh->length : BN_num_bits(dh->params.p) - 1;
+                l -= 2;
```

Comment:
Please explain/justify the `-2` here.
```suggestion
                l -= 1;
```

Reply:
Do we use one of the FIPS methods in the FIPS module?  Or does the FIPS module use the code patched in this PR?
If the latter, is that allowed?

Assuming FIPS is not a problem, I'd like to see a comment above `l -= 2` that explains that we're ensuring that $`l`$ is chosen to be the largest exponent with $`2^l - 1 < q`$, which requires $`l`$ to be one less than the bit length of $`q`$, i.e. two less than the bit length of $`p`$.

Reply:
This is  whole code section is inside an ifndef FIPS_MODULE section.
The FIPS compliant code using the algorithm in 5.6.1.1.4 is done here:
https://github.com/openssl/openssl/blob/49f8db53274191987b57d8e5542218690a983e35/crypto/ffc/ffc_key_generate.c#L12-L23
also the RFC 7919 groups are handled there, however while the RFC suggests to choose a
secret exponent in the range $[2^{N-1},2^{N}-1]$ the range is actually $[1,2^{N}-1]$